File: workers/price_series.py
"""Compact encoding for per-symbol daily price series.

One row per symbol rather than one row per symbol-day. The access pattern is
"draw one stock's chart", which reads a whole series and never filters inside
it, so a row-per-day table would cost ~250-400 MB and an index scan to answer a
question a single row answers. Cross-sectional questions are already served by
`screener_history`.

Three arrays per symbol, each delta-encoded then written as compact JSON:

* ``session_deltas`` -- gaps in the shared trading calendar. A symbol that
  traded every session encodes as ``[start, 1, 1, 1, ...]``, which compresses to
  almost nothing, and a thin stock's real gaps survive instead of being
  forward-filled into a flat line that never happened.
* ``closes`` -- adjusted close in paise, so the values are integers and the
  deltas are small.
* ``volumes`` -- shares traded.

Delta encoding is what makes JSON competitive with binary here: a price series
is smooth, so successive differences are one or two digits where absolute values
are six. Measured on a representative 2116-point series, close plus volume is
~22 KB raw and ~10 KB over HTTP gzip, against ~16 KB raw for float32 binary --
and JSON needs no endianness contract between Python and the browser.

Paise are exact: `round(close * 100)` on a two-decimal quote loses nothing, and
integer deltas cannot accumulate the float drift that a running sum of decimals
would.
"""

from __future__ import annotations

import json
from datetime import date, datetime

PAISE = 100


def _as_date(value):
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    return datetime.fromisoformat(str(value)[:10]).date()


def encode_deltas(values) -> str:
    """Delta-encode integers and serialise them as compact JSON.

    The first element is absolute; every later element is its difference from
    the one before. An empty series encodes as ``"[]"`` rather than raising, so
    a symbol with no observations is still representable.
    """
    numbers = [int(value) for value in values]
    if not numbers:
        return "[]"
    out = [numbers[0]]
    previous = numbers[0]
    for number in numbers[1:]:
        out.append(number - previous)
        previous = number
    return json.dumps(out, separators=(",", ":"))


def decode_deltas(text) -> list[int]:
    """Inverse of :func:`encode_deltas`."""
    if not text:
        return []
    numbers = json.loads(text)
    if not numbers:
        return []
    out = [int(numbers[0])]
    for delta in numbers[1:]:
        out.append(out[-1] + int(delta))
    return out


def build_series(sessions, observations):
    """Encode one symbol's series against the shared calendar.

    ``sessions`` is the ascending calendar. ``observations`` maps a session date
    to ``(adjusted_close, volume)``. Sessions the symbol did not trade are
    simply absent -- they are recorded as a gap in ``session_deltas``, never
    invented.

    Returns ``None`` when nothing usable is left, so a caller can skip a symbol
    instead of publishing an empty chart.
    """
    index_of = {day: position for position, day in enumerate(sessions)}
    indices, closes, volumes = [], [], []
    for day in sorted(observations):
        position = index_of.get(_as_date(day))
        if position is None:
            # Traded on a date the calendar does not contain. Dropping it keeps
            # the three arrays index-aligned, which every consumer relies on.
            continue
        close, volume = observations[day]
        if close is None or close <= 0:
            continue
        indices.append(position)
        closes.append(round(float(close) * PAISE))
        volumes.append(int(volume) if volume and volume > 0 else 0)

    if len(indices) < 2:
        return None

    return {
        "session_deltas": encode_deltas(indices),
        "closes": encode_deltas(closes),
        "volumes": encode_deltas(volumes),
        "points": len(indices),
        "first_session": sessions[indices[0]].isoformat(),
        "last_session": sessions[indices[-1]].isoformat(),
    }


def decode_series(row, sessions):
    """Expand an encoded row back into ``[{date, close, volume}, ...]``.

    Exists so the Python side can assert a round trip against exactly what the
    browser will reconstruct; the dashboard has its own copy of this logic.
    """
    indices = decode_deltas(row["session_deltas"])
    closes = decode_deltas(row["closes"])
    volumes = decode_deltas(row["volumes"])
    if not (len(indices) == len(closes) == len(volumes)):
        raise ValueError(
            f"misaligned series: {len(indices)} sessions, {len(closes)} closes, "
            f"{len(volumes)} volumes"
        )
    return [
        {
            "date": sessions[position].isoformat(),
            "close": close / PAISE,
            "volume": volume,
        }
        for position, close, volume in zip(indices, closes, volumes)
    ]


def encode_calendar(sessions) -> str:
    """The shared session calendar, stored once rather than per symbol."""
    return json.dumps(
        [_as_date(day).isoformat() for day in sessions], separators=(",", ":")
    )


def decode_calendar(text):
    return [_as_date(value) for value in json.loads(text)]

File: dashboard/lib/price-series.mjs
/**
 * Reader for the encoded daily price series.
 *
 * The writer is `workers/price_series.py`; this is the other half of that
 * contract and the two must agree exactly. Three delta-encoded JSON arrays per
 * symbol -- positions in the shared trading calendar, adjusted close in paise,
 * and volume -- chosen because a price series is smooth, so successive
 * differences are one or two digits where absolute values are six.
 *
 * A gap in `sessionDeltas` is a session the symbol did not trade. It is left as
 * a gap rather than forward-filled: a thin stock that traded eleven times in a
 * year should look like eleven observations, not a flat line.
 */

/** Paise per rupee. Prices are stored as integers to avoid float drift. */
const PAISE = 100;

/**
 * Undo delta encoding.
 *
 * @param {string | null | undefined} text
 * @returns {number[]}
 */
export function decodeDeltas(text) {
  if (!text) return [];
  const numbers = JSON.parse(text);
  if (!Array.isArray(numbers) || numbers.length === 0) return [];
  const out = new Array(numbers.length);
  out[0] = numbers[0];
  for (let index = 1; index < numbers.length; index += 1) {
    out[index] = out[index - 1] + numbers[index];
  }
  return out;
}

/**
 * Expand an encoded row into ascending `{ time, close, volume }` points.
 *
 * `time` is a plain `YYYY-MM-DD` string, which is what lightweight-charts wants
 * for a daily series -- passing epoch seconds would re-introduce the timezone
 * question that a daily bar does not have.
 *
 * Returns `[]` rather than throwing on a malformed row, because a broken chart
 * must not take the rest of the stock page down with it. Misalignment is
 * reported through `ok` so a caller can surface it.
 *
 * @param {{session_deltas: string, closes: string, volumes: string} | null | undefined} row
 * @param {string[]} sessions ISO dates, ascending
 * @returns {{ok: boolean, points: {time: string, close: number, volume: number}[]}}
 */
export function decodeSeries(row, sessions) {
  if (!row || !sessions?.length) return { ok: false, points: [] };

  const indices = decodeDeltas(row.session_deltas);
  const closes = decodeDeltas(row.closes);
  const volumes = decodeDeltas(row.volumes);

  if (indices.length !== closes.length || indices.length !== volumes.length) {
    return { ok: false, points: [] };
  }

  const points = [];
  for (let index = 0; index < indices.length; index += 1) {
    const day = sessions[indices[index]];
    // A position past the end of the calendar means the series was published
    // against a newer calendar than the one read. Skip rather than plotting the
    // point at `undefined`, which lightweight-charts renders at the epoch.
    if (day === undefined) continue;
    points.push({
      time: day,
      close: closes[index] / PAISE,
      volume: volumes[index],
    });
  }
  return { ok: true, points };
}

/**
 * Simple moving average over `close`, aligned to the same points.
 *
 * Emits nothing until the window is full. A 200-day average computed from 30
 * observations is not a 200-day average, and drawing it would put a confident
 * line in exactly the region where there is least information.
 *
 * @param {{time: string, close: number}[]} points
 * @param {number} window
 * @returns {{time: string, value: number}[]}
 */
export function movingAverage(points, window) {
  if (!points?.length || window < 1) return [];
  const out = [];
  let sum = 0;
  for (let index = 0; index < points.length; index += 1) {
    sum += points[index].close;
    if (index >= window) sum -= points[index - window].close;
    if (index >= window - 1) {
      out.push({ time: points[index].time, value: sum / window });
    }
  }
  return out;
}

/** Range buttons, in trading sessions. `null` means the whole series. */
export const RANGES = [
  { label: "1M", sessions: 21 },
  { label: "6M", sessions: 126 },
  { label: "1Y", sessions: 252 },
  { label: "3Y", sessions: 756 },
  { label: "5Y", sessions: 1260 },
  { label: "Max", sessions: null },
];

/**
 * The trailing slice a range button selects.
 *
 * Moving averages are computed on the *full* series and sliced afterwards, so
 * a 1M view still shows a true 200-day average rather than one restarted from
 * the left edge of the window. That is the difference between an average and a
 * line that merely looks like one.
 *
 * @template {{time: string}} T
 * @param {T[]} points
 * @param {number | null} sessions
 * @returns {T[]}
 */
export function sliceRange(points, sessions) {
  if (!points?.length) return [];
  if (!sessions || sessions >= points.length) return points;
  return points.slice(points.length - sessions);
}

/**
 * Trim a series to the visible window, keeping only what overlaps it.
 *
 * @template {{time: string}} T
 * @param {T[]} series
 * @param {string | undefined} from ISO date of the first visible point
 * @returns {T[]}
 */
export function clipFrom(series, from) {
  if (!from) return series;
  return series.filter((point) => point.time >= from);
}
